File: packages/cli/src/commands/update.ts
import { Command } from 'commander';
import * as path from 'node:path';
import * as os from 'node:os';
import { mkdir, writeFile } from 'node:fs/promises';
import {
  icons,
  c,
  log,
  logVerbose,
  logJson,
  withSpinner,
  setOutputMode,
  getOutputMode,
} from '../lib/output.js';
import { loadConfig } from '../lib/config.js';
import { createApiClient } from '../lib/api-client.js';
import { resolveSkills, parseSpecifier } from '../services/resolver.js';
import type { ResolvedSkill } from '../services/resolver.js';
import { linkSkill } from '../services/linker.js';
import {
  loadSkillsJson,
  loadLockFile,
  updateLockFile,
  getGlobalSkillsDir,
} from '../services/skills-json.js';
import { ensureSpmDirs } from '../services/preflight.js';

interface UpdateOpts {
  global?: boolean;
  verbose?: boolean;
  json?: boolean;
  silent?: boolean;
  dryRun?: boolean;
}

interface UpdateCandidate {
  name: string;
  currentVersion: string;
  newVersion: string;
  resolved: ResolvedSkill;
}

/**
 * Get the cache directory for a specific skill version.
 */
const getCacheDir = (name: string, version: string): string => {
  return path.join(os.homedir(), '.spm', 'cache', name, version);
};

/**
 * Get the extraction directory for a specific skill version.
 */
const getSkillDir = (name: string, version: string): string => {
  return path.join(os.homedir(), '.spm', 'skills', name, version);
};

/**
 * Download and cache a skill package.
 */
const downloadAndCache = async (
  apiClient: ReturnType<typeof createApiClient>,
  skill: ResolvedSkill,
): Promise<string> => {
  const cacheDir = getCacheDir(skill.name, skill.version);
  await mkdir(cacheDir, { recursive: true });

  const data = await apiClient.downloadSkill(skill.name, skill.version);
  const sklPath = path.join(cacheDir, `${skill.name}-${skill.version}.skl`);
  await writeFile(sklPath, Buffer.from(data));

  return sklPath;
};

/**
 * Extract a .skl file (tar.gz) to the skill store.
 */
const extractSkill = async (sklPath: string, name: string, version: string): Promise<string> => {
  const skillDir = getSkillDir(name, version);
  await mkdir(skillDir, { recursive: true });

  const { execFile } = await import('node:child_process');
  const { promisify } = await import('node:util');
  const execFileAsync = promisify(execFile);

  await execFileAsync('tar', ['-xzf', sklPath, '-C', skillDir]);
  return skillDir;
};

export const registerUpdateCommand = (program: Command): void => {
  program
    .command('update [skills...]')
    .aliases(['up', 'upgrade'])
    .description('Update skills to latest matching versions')
    .option('-g, --global', 'Update global skills')
    .option('--dry-run', 'Show what would be updated')
    .option('--verbose', 'Verbose output')
    .option('--json', 'JSON output')
    .option('--silent', 'No output')
    .action(async (skillNames: string[], opts: UpdateOpts) => {
      setOutputMode(getOutputMode(opts));
      await ensureSpmDirs();

      const isGlobal = opts.global ?? false;
      const targetDir = isGlobal ? getGlobalSkillsDir() : process.cwd();

      const skillsJson = await loadSkillsJson(targetDir);
      if (!skillsJson) {
        log(`${icons.info} No skills.json found. Nothing to update.`);
        return;
      }

      const lockFile = await loadLockFile(targetDir);

      // Determine which skills to check for updates
      let skillsToCheck: Array<{ name: string; range: string }>;
      if (skillNames.length > 0) {
        skillsToCheck = skillNames.map((input) => {
          const spec = parseSpecifier(input);
          const range = spec.range ?? skillsJson.skills[spec.name] ?? 'latest';
          return { name: spec.name, range };
        });
      } else {
        skillsToCheck = Object.entries(skillsJson.skills).map(([name, range]) => ({
          name,
          range,
        }));
      }

      if (skillsToCheck.length === 0) {
        log(`${icons.info} No skills to update.`);
        return;
      }

      const config = loadConfig();
      const apiClient = createApiClient({ registry: config.registry, token: config.token });

      // Resolve latest versions
      const { resolved } = await withSpinner('Checking for updates...', () =>
        resolveSkills(apiClient, skillsToCheck),
      );

      // Compare against lock file to find actual updates
      const candidates: UpdateCandidate[] = [];

      for (const skill of resolved) {
        const currentVersion = lockFile?.skills[skill.name]?.version;
        if (!currentVersion || currentVersion !== skill.version) {
          candidates.push({
            name: skill.name,
            currentVersion: currentVersion ?? 'not installed',
            newVersion: skill.version,
            resolved: skill,
          });
        }
      }

      if (candidates.length === 0) {
        log(`${icons.success} All skills are up to date.`);
        if (opts.json) {
          logJson({ command: 'update', status: 'success', updates: [] });
        }
        return;
      }

      // Show update plan
      log('');
      log(c.bold('Updates available:'));
      for (const candidate of candidates) {
        log(
          `  ${c.name(candidate.name)}: ${c.version(candidate.currentVersion)} ${icons.arrow} ${c.version(candidate.newVersion)}`,
        );
      }

      if (opts.dryRun) {
        log('');
        log(c.dim('Dry run — no changes made.'));
        if (opts.json) {
          logJson({
            command: 'update',
            status: 'dry-run',
            updates: candidates.map((cand) => ({
              name: cand.name,
              from: cand.currentVersion,
              to: cand.newVersion,
            })),
          });
        }
        return;
      }

      // Perform updates
      log('');
      for (const candidate of candidates) {
        const skill = candidate.resolved;

        const sklPath = await withSpinner(`Updating ${skill.name}@${skill.version}...`, () =>
          downloadAndCache(apiClient, skill),
        );

        const skillDir = await extractSkill(sklPath, skill.name, skill.version);
        const linkResult = await linkSkill(skillDir, skill.name);

        log(
          `  ${icons.success} ${c.name(skill.name)}: ${c.version(candidate.currentVersion)} ${icons.arrow} ${c.version(skill.version)}`,
        );
        logVerbose(`    Linked to: ${linkResult.agents.join(', ')}`);
      }

      // Update lock file
      await updateLockFile(
        targetDir,
        candidates.map((cand) => cand.resolved),
      );

      if (opts.json) {
        logJson({
          command: 'update',
          status: 'success',
          updates: candidates.map((cand) => ({
            name: cand.name,
            from: cand.currentVersion,
            to: cand.newVersion,
          })),
        });
        return;
      }

      log('');
      log(
        `${icons.success} ${candidates.length} skill${candidates.length === 1 ? '' : 's'} updated`,
      );
    });
};

File: cli-go/cmd/update.go
package cmd

import (
	"fmt"
	"os"
	"runtime"
	"strings"

	"github.com/skillpkg/spm/internal/api"
	"github.com/skillpkg/spm/internal/config"
	"github.com/skillpkg/spm/internal/output"
	"github.com/skillpkg/spm/internal/resolver"
	"github.com/skillpkg/spm/internal/skillsjson"
	"github.com/spf13/cobra"
)

var updateCmd = &cobra.Command{
	Use:     "update [name...]",
	Aliases: []string{"up", "upgrade"},
	Short:   "Update skills to latest matching versions",
	Long: `Check for newer versions of installed skills and update them.

If no arguments are given, updates all skills in skills.json.

Examples:
  spm update
  spm update my-skill
  spm update skill-a skill-b`,
	RunE: runUpdate,
}

func init() {
	rootCmd.AddCommand(updateCmd)
}

// updateEntry tracks an individual skill update.
type updateEntry struct {
	Name       string `json:"name"`
	OldVersion string `json:"from"`
	NewVersion string `json:"to"`
}

// updateJSONOutput is the JSON mode output for the update command.
type updateJSONOutput struct {
	Command string        `json:"command"`
	Status  string        `json:"status"`
	Updates []updateEntry `json:"updates"`
}

func runUpdate(cmd *cobra.Command, args []string) error {
	spmHome, err := config.HomeDir()
	if err != nil {
		return fmt.Errorf("determining SPM home: %w", err)
	}

	cwd, err := os.Getwd()
	if err != nil {
		return fmt.Errorf("getting working directory: %w", err)
	}

	// Load skills.json
	sj, err := skillsjson.LoadSkillsJson(cwd)
	if err != nil {
		return fmt.Errorf("loading skills.json: %w", err)
	}
	if sj == nil || len(sj.Skills) == 0 {
		Out.Log("%s No skills.json found. Nothing to update.", output.Icons["info"])
		return nil
	}

	// Load lock file to know current versions
	lock, err := skillsjson.LoadLockFile(cwd)
	if err != nil {
		Out.LogVerbose("Failed to load lock file: %s", err)
	}

	// Determine which skills to check
	type checkItem struct {
		name   string
		vrange string
	}
	var toCheck []checkItem

	if len(args) > 0 {
		for _, arg := range args {
			spec, err := resolver.Parse(arg)
			if err != nil {
				return fmt.Errorf("invalid specifier %q: %w", arg, err)
			}
			name := spec.FullName()
			vrange := spec.VersionRange
			if vrange == "" {
				// Use range from skills.json if available
				if r, ok := sj.Skills[name]; ok {
					vrange = r
				} else {
					vrange = "latest"
				}
			}
			toCheck = append(toCheck, checkItem{name: name, vrange: vrange})
		}
	} else {
		for name, vrange := range sj.Skills {
			toCheck = append(toCheck, checkItem{name: name, vrange: vrange})
		}
	}

	if len(toCheck) == 0 {
		Out.Log("%s No skills to update.", output.Icons["info"])
		return nil
	}

	// Resolve latest versions
	resolveSkills := make([]api.ResolveSkill, 0, len(toCheck))
	for _, item := range toCheck {
		resolveSkills = append(resolveSkills, api.ResolveSkill{
			Name:  item.name,
			Range: item.vrange,
		})
	}

	client := api.NewClient(Cfg.RegistryURL(), Cfg.Token)

	sp := Out.StartSpinner("Checking for updates...")
	resolveResp, err := client.Resolve(resolveSkills, runtime.GOOS)
	output.StopSpinner(sp)
	if err != nil {
		return fmt.Errorf("resolving skills: %w", err)
	}

	// Report unresolved
	for _, u := range resolveResp.Unresolved {
		Out.LogError("Could not resolve: %s - %s", u.Name, u.Error)
	}

	// Compare resolved versions against lock file
	var candidates []struct {
		resolved   api.ResolvedSkill
		oldVersion string
		newVersion string
	}

	for _, skill := range resolveResp.Resolved {
		oldVersion := "not installed"
		if lock != nil {
			if entry, ok := lock.Skills[skill.Name]; ok {
				oldVersion = entry.Version
				if entry.Version == skill.Version {
					continue // already up to date
				}
			}
		}
		candidates = append(candidates, struct {
			resolved   api.ResolvedSkill
			oldVersion string
			newVersion string
		}{
			resolved:   skill,
			oldVersion: oldVersion,
			newVersion: skill.Version,
		})
	}

	if len(candidates) == 0 {
		msg := fmt.Sprintf("%s All skills are up to date.", output.Icons["success"])
		Out.Log(msg)
		if Out.Mode == output.ModeJSON {
			return Out.LogJSON(updateJSONOutput{
				Command: "update",
				Status:  "success",
				Updates: []updateEntry{},
			})
		}
		return nil
	}

	// Show update plan
	Out.Log("")
	Out.Log("%s Updates available:", output.Bold(""))
	for _, c := range candidates {
		Out.Log("  %s: %s %s %s",
			output.Cyan(c.resolved.Name),
			output.Dim(c.oldVersion),
			output.Icons["arrow"],
			output.Green(c.newVersion))
	}

	// Perform updates
	Out.Log("")
	updates := make([]updateEntry, 0, len(candidates))
	lockUpdates := make([]skillsjson.ResolvedSkill, 0, len(candidates))

	for _, c := range candidates {
		skill := c.resolved

		res, err := downloadExtractLink(client, skill, spmHome, true)
		if err != nil {
			Out.LogError("Failed to update %s: %s", skill.Name, err)
			continue
		}

		updates = append(updates, updateEntry{
			Name:       skill.Name,
			OldVersion: c.oldVersion,
			NewVersion: c.newVersion,
		})

		lockUpdates = append(lockUpdates, skillsjson.ResolvedSkill{
			Name:        skill.Name,
			Version:     skill.Version,
			DownloadURL: skill.DownloadURL,
			Checksum:    skill.ChecksumSHA256,
		})

		Out.Log("  %s %s: %s %s %s",
			output.Icons["success"],
			output.Cyan(skill.Name),
			output.Dim(c.oldVersion),
			output.Icons["arrow"],
			output.Green(c.newVersion))

		if len(res.LinkedAgents) > 0 {
			Out.LogVerbose("Linked to: %s", strings.Join(res.LinkedAgents, ", "))
		}
	}

	// Update lock file
	if len(lockUpdates) > 0 {
		if err := skillsjson.UpdateLockFile(cwd, lockUpdates); err != nil {
			Out.LogVerbose("Failed to update lock file: %s", err)
		}
	}

	if Out.Mode == output.ModeJSON {
		return Out.LogJSON(updateJSONOutput{
			Command: "update",
			Status:  "success",
			Updates: updates,
		})
	}

	Out.Log("")
	Out.Log("%s %d skill(s) updated", output.Icons["success"], len(updates))
	return nil
}
